File: zig/3-pid-bm/main/Services.zig
const std = @import("std");
const idf = @import("esp_idf");
const sys = idf.sys;

const Wifi = @import("Wifi.zig");
const c = @import("c.zig");

const Self = @This();

const mdns_port = 502;
const mdns_hostname = "esp32";

mdns: MDns,
wifi: Wifi,

pub fn init() !Self {
    idf.espCheckError(sys.nvs_flash_init()) catch |err| {
        switch (err) {
            idf.esp_error.ErrorNvsNoFreePages,
            idf.esp_error.ErrorNvsNewVersionFound,
            => {
                try idf.espCheckError(sys.nvs_flash_erase());
                try idf.espCheckError(sys.nvs_flash_init());
            },
            else => return err,
        }
    };
    errdefer idf.espLogError(sys.nvs_flash_deinit());

    try idf.espCheckError(sys.esp_event_loop_create_default());
    errdefer idf.espLogError(sys.esp_event_loop_delete_default());

    const mdns = try MDns.init();
    errdefer mdns.deinit();

    const wifi = try Wifi.init();
    errdefer wifi.deinit();

    return .{
        .mdns = mdns,
        .wifi = wifi,
    };
}

pub fn deinit(self: *const Self) void {
    self.wifi.deinit();
    self.mdns.deinit();
    idf.espLogError(sys.esp_event_loop_delete_default());
    idf.espLogError(sys.nvs_flash_deinit());
}

const MDns = struct {
    fn init() !MDns {
        try c.espCheckError(c.mdns_init());
        errdefer c.mdns_free();

        try c.espCheckError(c.mdns_hostname_set(mdns_hostname));

        var txt_items = [_]c.mdns_txt_item_t{
            .{ .key = "board", .value = "esp32" },
        };
        const res = c.mdns_service_add(mdns_hostname, "_modbus", "_tcp", mdns_port, &txt_items, txt_items.len);
        try c.espCheckError(res);

        return .{};
    }

    fn deinit(_: MDns) void {
        c.mdns_free();
    }
};

File: c/3-pid-bm/main/services.c
#include "esp_err.h"
#include "esp_event.h"
#include "esp_log.h"
#include "mdns.h"
#include "nvs_flash.h"

#include "services.h"

#define MB_MDNS_PORT (502)
#define MB_MDNS_HOSTNAME "mst"

static const char TAG[] = "services";

esp_err_t my_mdns_init(void) {
  esp_err_t err;

  err = mdns_init();
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "mdns_init fail (0x%x)", (int)err);
    return err;
  }

  err = mdns_hostname_set(MB_MDNS_HOSTNAME);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "mdns_hostname_set fail (0x%x)", (int)err);
    mdns_free();
    return err;
  }
  ESP_LOGI(TAG, "mdns hostname set to: [%s]", MB_MDNS_HOSTNAME);

  mdns_txt_item_t serviceTxtData[] = {{"board", "esp32"}};

  err = mdns_service_add(
      MB_MDNS_HOSTNAME, "_modbus", "_tcp", MB_MDNS_PORT, serviceTxtData, 1
  );
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "mdns_service_add fail (0x%x)", (int)err);
    mdns_free();
    return err;
  }

  return ESP_OK;
}

void my_mdns_deinit(void) { mdns_free(); }

esp_err_t services_init(services_t *self) {
  esp_err_t err = nvs_flash_init();
  if (err == ESP_ERR_NVS_NO_FREE_PAGES ||
      err == ESP_ERR_NVS_NEW_VERSION_FOUND) {
    err = nvs_flash_erase();
    if (err != ESP_OK) {
      ESP_LOGE(TAG, "nvs_flash_erase (0x%x)", (int)err);
      return err;
    }
    err = nvs_flash_init();
  }
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "nvs_flash_init fail (0x%x)", (int)err);
    return err;
  }

  err = esp_event_loop_create_default();
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "esp_event_loop_create_default fail (0x%x)", (int)err);
    ESP_ERROR_CHECK_WITHOUT_ABORT(nvs_flash_deinit());
    return err;
  }

  err = my_mdns_init();
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "my_mdns_init fail (0x%x)", (int)err);
    ESP_ERROR_CHECK_WITHOUT_ABORT(esp_event_loop_delete_default());
    ESP_ERROR_CHECK_WITHOUT_ABORT(nvs_flash_deinit());
    return err;
  }

  err = my_wifi_init(&self->wifi);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "my_wifi_init fail (0x%x)", (int)err);
    my_mdns_deinit();
    ESP_ERROR_CHECK_WITHOUT_ABORT(esp_event_loop_delete_default());
    ESP_ERROR_CHECK_WITHOUT_ABORT(nvs_flash_deinit());
    return err;
  }

  return ESP_OK;
}

void services_deinit(services_t *self) {
  my_wifi_deinit(&self->wifi);
  ESP_ERROR_CHECK_WITHOUT_ABORT(esp_event_loop_delete_default());
  ESP_ERROR_CHECK_WITHOUT_ABORT(esp_netif_deinit());
  ESP_ERROR_CHECK_WITHOUT_ABORT(nvs_flash_deinit());
  my_mdns_deinit();
}
